pkg/apis/etcd/v1beta2/backup_types.go

```diff
@@ -47,8 +47,8 @@ type EtcdBackup struct {
 
 // BackupSpec contains a backup specification for an etcd cluster.
 type BackupSpec struct {
-	// ClusterName is the etcd cluster name.
-	ClusterName string `json:"clusterName,omitempty"`
+	// EtcdEndpoints specifies the etcd cluster endpoints.
```

Comment:
let us document how will the endpoint be used. will the snapshot come from the first available endpoint? or it will get from the endpoints with the largest revision? can i put endpoints of different clusters? what happens if i do?

Reply:
talking to hongqiao. it seems like we just pass endpoints to etcd client and let etcd client decide which endpoint to get etcd snapshot from.

Reply:
that changes the previous logic. the previous logic tries to snapshot from the server that has the largest revision. 

why do we want to make it less smarter?

Reply:
@xiang90 should we keep the getting snapshot logic as simple as possible first? 

user can still achieve saving the snapshot with max revision by giving an endpoint that points the max member.

Reply:
how will the user be able to do that? i do not think the goal is to make the logic as simple as possible. the goal is to decouple the backup-operator from the etcd-operator logic by not taking the cluster name.

Reply:
Here's the use case I have in mind.
Instead of getting all transient etcd pods' IP, initially user will give the load balancer URL to us. It doesn't make sense to retrieve the max revision.

Reply:
> Instead of getting all transient etcd pods' IP, initially user will give the load balancer URL to us. It doesn't make sense to retrieve the max revision.

If that is the case, the snapshot logic has to be changed. we need to:

1. do a l-get to ensure the member we are interacting with is up-to-date
2. do a snapshot

whatever we choose to do here, we should not do that in the PR. we should first finish the decouple work.

Reply:
The issue is that if user give us a load balancer URL, then there is a race condition that you first get a revision of X, but then you get a snapshot of Y.

Reply:
> The issue is that if user give us a load balancer URL, then there is a race condition that you first get a revision of X, but then you get a snapshot of Y.

no. that is not how etcd client works. the client wont change connection between requests.

Reply:
Cool. It seems I have made the wrong assumption.
So it's fine then.

Reply:
`how will the user be able to do that? `
I am thinking that user manually figure out which member has highest rev using etcdctl, but that seems tedious.

Reply:
> I am thinking that user manually figure out which member has highest rev using etcdctl, but that seems tedious.

right. then i would rather to request the snapshot with etcdctl and pipe that into s3util to upload :P

Reply:
Actually the below code has the race I mentioned above..
So needs to fix that too.

Reply:
> Actually the below code has the race I mentioned above..

just to be clear, the goal of the sorting is to not get a super out of date snapshot rather than trying to ensure we get the most recent snapshot. an endpoint with smaller revision can always increase its revision faster than the one was the largest. we do not need to handle that race, and we cannot.